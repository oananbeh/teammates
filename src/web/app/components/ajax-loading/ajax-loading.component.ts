import { Component, Input } from '@angular/core';

/**
 * Displaying the ajax loader.
 */
@Component({
  selector: 'tm-ajax-loading',
  templateUrl: './ajax-loading.component.html',
  styleUrls: ['./ajax-loading.component.scss'],
})
export class AjaxLoadingComponent {

  @Input()
  useBlueSpinner: boolean = false;

  @Input()
  color: string = 'white';

}
