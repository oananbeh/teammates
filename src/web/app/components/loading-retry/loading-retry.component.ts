import { Component, EventEmitter, Input, Output } from '@angular/core';

/**
 * Retry button for when loading fails.
 */
@Component({
  selector: 'tm-loading-retry',
  templateUrl: './loading-retry.component.html',
  styleUrls: ['./loading-retry.component.scss'],
})
export class LoadingRetryComponent {

  @Input() message: string = '';

  @Input() shouldShowRetry: boolean = false;

  @Input() retryButtonDisabled: boolean = false;

  @Output() retryEvent: EventEmitter<any> = new EventEmitter<any>();

}
