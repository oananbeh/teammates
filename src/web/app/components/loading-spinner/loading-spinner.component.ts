import { Component } from '@angular/core';

/**
 * Loading spinner to show when waiting for request.
 */
@Component({
  selector: 'tm-loading-spinner',
  templateUrl: './loading-spinner.component.html',
  styleUrls: ['./loading-spinner.component.scss'],
})
export class LoadingSpinnerComponent {
  loadingMessage: string = 'Loading...';

  constructor() {
    // After 5 seconds of loading show message to assure user the page is not hanging.
    setTimeout(() => { this.loadingMessage = 'Still loading...'; }, 5000);
  }
}
