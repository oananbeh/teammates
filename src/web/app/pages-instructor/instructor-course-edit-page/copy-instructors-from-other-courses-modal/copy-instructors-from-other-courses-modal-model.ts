import { Instructor } from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';

/**
 * The model for a course tab.
 */
export interface CourseTabModel {
  courseId: string;
  courseName: string;
  creationTimestamp: number;
  isArchived: boolean;

  instructorCandidates: InstructorToCopyCandidateModel[];
  instructorCandidatesSortBy: SortBy;
  instructorCandidatesSortOrder: SortOrder;

  hasInstructorsLoaded: boolean;
  isTabExpanded: boolean;
  hasLoadingFailed: boolean;
}

/**
 * The model for an instructor to copy.
 */
export interface InstructorToCopyCandidateModel {
  instructor: Instructor;
  isSelected: boolean;
}
