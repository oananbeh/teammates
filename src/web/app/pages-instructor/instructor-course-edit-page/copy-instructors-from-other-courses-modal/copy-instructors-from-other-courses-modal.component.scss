.sortable-header {
  cursor: pointer;

  i {
    margin-left: 0.3em;
  }

  .fa-stack {
    width: 20%;
  }

  button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    text-align: left;
    width: 100%;
    font: inherit;
    cursor: inherit;
  }
}

#copy-instructor-modal {
  table-layout: auto;
  margin-bottom: 10px;
  min-width: 70vw;
}

.modal-footer {
  justify-content: right;
  background-color: white;
}

.form-control {
  height: calc(0.5em + 1.5vh);
  width: calc(0.5em + 1.5vw);
}

.no-shadow {
  box-shadow: none;
}
