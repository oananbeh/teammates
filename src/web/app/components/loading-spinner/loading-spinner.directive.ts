import {
  ComponentRef,
  Directive,
  Input,
  TemplateRef,
  ViewContainerRef,
} from '@angular/core';
import { LoadingSpinnerComponent } from './loading-spinner.component';

/**
 * Directive for loading spinner component
 */
@Directive({
  selector: '[tmIsLoading]',
})
export class LoadingSpinnerDirective {
  loadingComponent: ComponentRef<LoadingSpinnerComponent>;

  constructor(private templateRef: TemplateRef<any>,
              private viewContainer: ViewContainerRef) {
    this.loadingComponent = this.viewContainer.createComponent(LoadingSpinnerComponent);
  }

  @Input() set tmIsLoading(loading: boolean) {
    this.viewContainer.clear();

    if (loading) {
      this.loadingComponent = this.viewContainer.createComponent(LoadingSpinnerComponent);
    } else {
      this.viewContainer.createEmbeddedView(this.templateRef);
    }
  }
}
