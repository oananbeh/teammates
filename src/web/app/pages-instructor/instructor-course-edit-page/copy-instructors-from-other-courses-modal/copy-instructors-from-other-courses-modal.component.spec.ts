import { CommonModule } from '@angular/common';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule } from '@angular/forms';
import { NgbActiveModal, NgbTooltipModule } from '@ng-bootstrap/ng-bootstrap';
import { of } from 'rxjs';
import { CourseTabModel, InstructorToCopyCandidateModel } from './copy-instructors-from-other-courses-modal-model';
import { CopyInstructorsFromOtherCoursesModalComponent } from './copy-instructors-from-other-courses-modal.component';
import { InstructorService } from '../../../../services/instructor.service';
import {
  Instructor,
  InstructorPermissionRole, Instructors,
  JoinState,
} from '../../../../types/api-output';
import { SortBy, SortOrder } from '../../../../types/sort-properties';
import { AjaxLoadingModule } from '../../../components/ajax-loading/ajax-loading.module';
import { LoadingRetryModule } from '../../../components/loading-retry/loading-retry.module';
import { LoadingSpinnerModule } from '../../../components/loading-spinner/loading-spinner.module';
import { PanelChevronModule } from '../../../components/panel-chevron/panel-chevron.module';
import { TeammatesCommonModule } from '../../../components/teammates-common/teammates-common.module';

describe('CopyInstructorsFromOtherCoursesModalComponent', () => {

  const testInstructor1: Instructor = {
    googleId: 'googleIdOfIns1',
    courseId: 'FAN0001',
    email: 'ins1@fan.tmt',
    isDisplayedToStudents: true,
    displayedToStudentsAs: 'abcdefg',
    name: 'Instructor Cat',
    role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_COOWNER,
    joinState: JoinState.JOINED,
    key: 'ujg2l5hj5l7jrfo28fyosfklh2892',
    institute: 'A',
  };

  const testInstructor2: Instructor = {
    googleId: 'googleIdOfIns2',
    courseId: 'FAN0001',
    email: 'ins2@fan.tmt',
    isDisplayedToStudents: false,
    displayedToStudentsAs: '',
    name: 'Instructor Dog',
    role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_CUSTOM,
    joinState: JoinState.JOINED,
    key: '1343jbcl3iru2yct0897goji',
    institute: 'A',
  };

  const testInstructor3: Instructor = {
    googleId: 'googleIdOfIns3',
    courseId: 'FAN0002',
    email: 'ins1@fan.tmt',
    isDisplayedToStudents: true,
    displayedToStudentsAs: 'QWQWQWQ',
    name: 'King',
    role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_MANAGER,
    joinState: JoinState.NOT_JOINED,
    key: 'pjkjbnc523iosdk2389nfbfib2',
    institute: 'A',
  };

  const testInstructor4: Instructor = {
    googleId: 'googleIdOfIns4',
    courseId: 'FAN0002',
    email: 'ins4@fan.tmt',
    isDisplayedToStudents: false,
    displayedToStudentsAs: 'oldname',
    name: 'Nothing',
    role: InstructorPermissionRole.INSTRUCTOR_PERMISSION_ROLE_TUTOR,
    joinState: JoinState.JOINED,
    key: 'qdnjvbi47t928cjqnk1o8ry2o',
    institute: 'A',
  };

  const testInstructorCandidate1: InstructorToCopyCandidateModel = {
    instructor: testInstructor1,
    isSelected: false,
  };

  const testInstructorCandidate2: InstructorToCopyCandidateModel = {
    instructor: testInstructor2,
    isSelected: false,
  };

  const testInstructorCandidate3: InstructorToCopyCandidateModel = {
    instructor: testInstructor3,
    isSelected: false,
  };

  const testInstructorCandidate4: InstructorToCopyCandidateModel = {
    instructor: testInstructor4,
    isSelected: false,
  };

  const testCourseTab1: CourseTabModel = {
    courseId: 'FAN0002',
    courseName: 'Test Course 1',
    creationTimestamp: new Date('2022-07-26T01:00:15Z').getTime(),
    isArchived: true,
    instructorCandidates: [],
    instructorCandidatesSortBy: SortBy.NONE,
    instructorCandidatesSortOrder: SortOrder.ASC,
    hasInstructorsLoaded: false,
    isTabExpanded: false,
    hasLoadingFailed: false,
  };

  const testCourseTab2: CourseTabModel = {
    courseId: 'FAN0001',
    courseName: 'Test Course 2',
    creationTimestamp: new Date('2022-02-22T22:22:22Z').getTime(),
    isArchived: false,
    instructorCandidates: [],
    instructorCandidatesSortBy: SortBy.NONE,
    instructorCandidatesSortOrder: SortOrder.ASC,
    hasInstructorsLoaded: false,
    isTabExpanded: false,
    hasLoadingFailed: false,
  };

  const testCourseTab3: CourseTabModel = {
    courseId: 'CS2103T',
    courseName: 'XXX Software Engineering',
    creationTimestamp: new Date('2022-06-21T07:51:20Z').getTime(),
    isArchived: false,
    instructorCandidates: [],
    instructorCandidatesSortBy: SortBy.NONE,
    instructorCandidatesSortOrder: SortOrder.ASC,
    hasInstructorsLoaded: false,
    isTabExpanded: false,
    hasLoadingFailed: false,
  };

  let component: CopyInstructorsFromOtherCoursesModalComponent;
  let fixture: ComponentFixture<CopyInstructorsFromOtherCoursesModalComponent>;
  let instructorService: InstructorService;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [CopyInstructorsFromOtherCoursesModalComponent],
      imports: [
        CommonModule,
        FormsModule,
        NgbTooltipModule,
        TeammatesCommonModule,
        PanelChevronModule,
        LoadingSpinnerModule,
        LoadingRetryModule,
        AjaxLoadingModule,
        HttpClientTestingModule,
      ],
      providers: [
        NgbActiveModal,
        InstructorService,
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(CopyInstructorsFromOtherCoursesModalComponent);
    instructorService = TestBed.inject(InstructorService);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should snap with courses', () => {
    component.courses = [testCourseTab1, testCourseTab2];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when instructors are loading', () => {
    testCourseTab1.isTabExpanded = true;
    testCourseTab1.hasLoadingFailed = false;
    testCourseTab1.hasInstructorsLoaded = false;
    testCourseTab1.instructorCandidates = [];
    component.courses = [testCourseTab1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when instructors are loaded', () => {
    testCourseTab1.isTabExpanded = true;
    testCourseTab1.hasLoadingFailed = false;
    testCourseTab1.hasInstructorsLoaded = true;
    testCourseTab1.instructorCandidates = [testInstructorCandidate3, testInstructorCandidate4];
    component.courses = [testCourseTab1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when feedback questions failed to load', () => {
    testCourseTab1.isTabExpanded = true;
    testCourseTab1.hasLoadingFailed = true;
    testCourseTab1.hasInstructorsLoaded = false;
    testCourseTab1.instructorCandidates = [];
    component.courses = [testCourseTab1];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should load instructors', () => {
    const instructors: Instructors = {
      instructors: [testInstructor1, testInstructor2],
    };
    jest.spyOn(instructorService, 'loadInstructors').mockReturnValue(of(instructors));
    component.courses = [testCourseTab2];

    component.loadInstructors(testCourseTab2);

    expect(component.courses[0].hasInstructorsLoaded).toBeTruthy();
    expect(component.courses[0].hasLoadingFailed).toBeFalsy();
    expect(component.courses[0].instructorCandidates.length).toBe(2);
    expect(component.courses[0].instructorCandidates[0].instructor.googleId)
      .toBe(testInstructor1.googleId);
    expect(component.courses[0].instructorCandidates[1].instructor.googleId)
      .toBe(testInstructor2.googleId);
  });

  it('should not allow copying when no instructors are selected', () => {
    testCourseTab2.instructorCandidates = [testInstructorCandidate1, testInstructorCandidate2];
    testCourseTab1.instructorCandidates = [testInstructorCandidate3, testInstructorCandidate4];
    component.courses = [testCourseTab2, testCourseTab1];
    fixture.detectChanges();

    const instructors: Instructor[] = component.getSelectedInstructors();
    const button: any = fixture.debugElement.nativeElement.querySelector('#btn-confirm-copy-instructor');

    expect(component.isAnyInstructorCandidatesSelected).toBeFalsy();
    expect(instructors.length).toBe(0);
    expect(button.disabled).toBeTruthy();
  });

  it('should trigger copy button clicked event with selected instructors', () => {
    testInstructorCandidate1.isSelected = true;
    testInstructorCandidate3.isSelected = true;
    testInstructorCandidate4.isSelected = true;
    testCourseTab2.instructorCandidates = [testInstructorCandidate1, testInstructorCandidate2];
    testCourseTab1.instructorCandidates = [testInstructorCandidate3, testInstructorCandidate4];
    component.courses = [testCourseTab2, testCourseTab1];
    fixture.detectChanges();

    const instructors: Instructor[] = [testInstructor1, testInstructor3, testInstructor4];
    const buttonConfirmCopy: any = fixture.debugElement.nativeElement.querySelector('#btn-confirm-copy-instructor');
    jest.spyOn(component.copyClickedEvent, 'emit');

    expect(component.isAnyInstructorCandidatesSelected).toBeTruthy();
    expect(buttonConfirmCopy.disabled).toBeFalsy();

    buttonConfirmCopy.click();
    expect(component.copyClickedEvent.emit).toHaveBeenCalledTimes(1);
    expect(component.copyClickedEvent.emit).toHaveBeenCalledWith(instructors);
  });

  it('should disable buttons when copying instructors', () => {
    testInstructorCandidate1.isSelected = true;
    testCourseTab2.instructorCandidates = [testInstructorCandidate1];
    component.courses = [testCourseTab2];
    fixture.detectChanges();

    const buttonConfirmCopy: any = fixture.debugElement.nativeElement.querySelector('#btn-confirm-copy-instructor');
    const buttonCancelCopy: any = fixture.debugElement.nativeElement.querySelector('#btn-cancel-copy-instructor');
    const buttonCloseModal: any = fixture.debugElement.nativeElement.querySelector('#btn-close-modal');

    buttonConfirmCopy.click();
    fixture.detectChanges();

    expect(buttonConfirmCopy.disabled).toBeTruthy();
    expect(buttonCancelCopy.disabled).toBeTruthy();
    expect(buttonCloseModal.disabled).toBeTruthy();
  });

  it('should sort courses by their IDs', () => {
    component.courses = [testCourseTab2, testCourseTab3, testCourseTab1];
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement.querySelector('#sort-course-id');
    button.click();
    expect(component.courses[0].courseId).toEqual('CS2103T');
    expect(component.courses[1].courseId).toEqual('FAN0001');
    expect(component.courses[2].courseId).toEqual('FAN0002');
  });

  it('should sort courses by their names', () => {
    component.courses = [testCourseTab2, testCourseTab3, testCourseTab1];
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement.querySelector('#sort-course-name');
    button.click();
    expect(component.courses[0].courseId).toEqual('FAN0002');
    expect(component.courses[1].courseId).toEqual('FAN0001');
    expect(component.courses[2].courseId).toEqual('CS2103T');
  });

  it('should sort courses by their creation dates', () => {
    component.courses = [testCourseTab2, testCourseTab3, testCourseTab1];
    fixture.detectChanges();

    const button: any = fixture.debugElement.nativeElement.querySelector('#sort-course-creation-date');
    button.click();
    expect(component.courses[0].courseId).toEqual('FAN0002');
    expect(component.courses[1].courseId).toEqual('CS2103T');
    expect(component.courses[2].courseId).toEqual('FAN0001');
  });
});
