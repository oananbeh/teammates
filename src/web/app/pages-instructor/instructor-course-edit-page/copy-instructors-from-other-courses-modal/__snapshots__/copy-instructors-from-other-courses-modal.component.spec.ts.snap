// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyInstructorsFromOtherCoursesModalComponent should snap when feedback questions failed to load 1`] = `
<tm-copy-instructors-from-other-courses-modal
  InstructorPermissionRole={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  copyClickedEvent={[Function EventEmitter_]}
  courses={[Function Array]}
  instructorService={[Function InstructorService]}
  isCopyingSelectedInstructors="false"
  notDisplayedToStudentText={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Instructors
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-modal"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-instructor-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Courses By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-course-name"
        >
          Course Name
        </button>
        <button
          class="btn btn-light"
          id="sort-course-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-info"
        >
          <strong
            class="text-break"
          >
            [FAN0002] (Archived): Test Course 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <div
              class="text-center"
            >
              <p
                class="text-muted m-1"
              >
                Failed to load instructors
              </p>
              <button
                class="btn btn-primary"
              >
                <i
                  aria-hidden="true"
                  class="fas fa-redo"
                />
                Â Retry
              </button>
            </div>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-instructor"
      type="button"
    >
      Copy 
    </button>
    <button
      class="btn btn-light"
      id="btn-cancel-copy-instructor"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-instructors-from-other-courses-modal>
`;

exports[`CopyInstructorsFromOtherCoursesModalComponent should snap when instructors are loaded 1`] = `
<tm-copy-instructors-from-other-courses-modal
  InstructorPermissionRole={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  copyClickedEvent={[Function EventEmitter_]}
  courses={[Function Array]}
  instructorService={[Function InstructorService]}
  isCopyingSelectedInstructors="false"
  notDisplayedToStudentText={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Instructors
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-modal"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-instructor-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Courses By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-course-name"
        >
          Course Name
        </button>
        <button
          class="btn btn-light"
          id="sort-course-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-info"
        >
          <strong
            class="text-break"
          >
            [FAN0002] (Archived): Test Course 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <div>
              <div
                class="card-body"
              >
                <table
                  class="table table-bordered table-responsive-lg"
                  id="copy-instructor-table"
                >
                  <thead
                    class="bg-info"
                  >
                    <tr>
                      <th>
                        #
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Instructor Name 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Instructor Email 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Displayed to Student As 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                      <th
                        aria-sort="none"
                        class="sortable-header"
                      >
                        <button>
                           Access Level 
                          <span
                            aria-hidden="true"
                            class="fa-stack"
                          >
                            <i
                              class="fas fa-sort"
                            />
                          </span>
                        </button>
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        <div>
                          <input
                            class="form-check-input no-shadow ng-untouched ng-pristine ng-valid"
                            id="enabled-checkbox"
                            type="checkbox"
                          />
                        </div>
                      </td>
                      <td>
                        King
                      </td>
                      <td>
                        ins1@fan.tmt
                      </td>
                      <td>
                        QWQWQWQ
                      </td>
                      <td>
                        Manager
                      </td>
                    </tr>
                    <tr>
                      <td>
                        <div>
                          <input
                            class="form-check-input no-shadow ng-untouched ng-pristine ng-valid"
                            id="enabled-checkbox"
                            type="checkbox"
                          />
                        </div>
                      </td>
                      <td>
                        Nothing
                      </td>
                      <td>
                        ins4@fan.tmt
                      </td>
                      <td>
                        (NOT displayed to students)
                      </td>
                      <td>
                        Tutor
                      </td>
                    </tr>
                  </tbody>
                </table>
                <button
                  class="btn btn-secondary float-end mt-2 mb-3"
                >
                  <i
                    aria-hidden="true"
                    class="fas fa-redo"
                  />
                   Â Reload 
                </button>
              </div>
            </div>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-instructor"
      type="button"
    >
      Copy 
    </button>
    <button
      class="btn btn-light"
      id="btn-cancel-copy-instructor"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-instructors-from-other-courses-modal>
`;

exports[`CopyInstructorsFromOtherCoursesModalComponent should snap when instructors are loading 1`] = `
<tm-copy-instructors-from-other-courses-modal
  InstructorPermissionRole={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  copyClickedEvent={[Function EventEmitter_]}
  courses={[Function Array]}
  instructorService={[Function InstructorService]}
  isCopyingSelectedInstructors="false"
  notDisplayedToStudentText={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Instructors
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-modal"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-instructor-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Courses By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-course-name"
        >
          Course Name
        </button>
        <button
          class="btn btn-light"
          id="sort-course-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-info"
        >
          <strong
            class="text-break"
          >
            [FAN0002] (Archived): Test Course 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="true"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-up"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
        <div>
          <tm-loading-retry>
            <tm-loading-spinner>
              <div
                class="loading-container"
              >
                <div
                  class="row spinner-border text-primary"
                  role="status"
                />
                <div
                  class="row text-secondary"
                >
                  Loading...
                </div>
              </div>
            </tm-loading-spinner>
          </tm-loading-retry>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-instructor"
      type="button"
    >
      Copy 
    </button>
    <button
      class="btn btn-light"
      id="btn-cancel-copy-instructor"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-instructors-from-other-courses-modal>
`;

exports[`CopyInstructorsFromOtherCoursesModalComponent should snap with courses 1`] = `
<tm-copy-instructors-from-other-courses-modal
  InstructorPermissionRole={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  copyClickedEvent={[Function EventEmitter_]}
  courses={[Function Array]}
  instructorService={[Function InstructorService]}
  isCopyingSelectedInstructors="false"
  notDisplayedToStudentText={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Instructors
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-modal"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-instructor-modal"
  >
    <div
      class="col-12 mb-3 text-md-end"
    >
      <strong
        class="d-inline"
      >
         Sort Courses By: 
      </strong>
      <div
        class="btn-group"
        data-toggle="buttons"
      >
        <button
          class="btn btn-light"
          id="sort-course-id"
        >
          Course ID
        </button>
        <button
          class="btn btn-light"
          id="sort-course-name"
        >
          Course Name
        </button>
        <button
          class="btn btn-light"
          id="sort-course-creation-date"
        >
          Creation Date
        </button>
      </div>
    </div>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-info"
        >
          <strong
            class="text-break"
          >
            [FAN0002] (Archived): Test Course 1
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="false"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-down"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="card"
      >
        <div
          class="card-header cursor-pointer bg-primary text-white"
        >
          <strong
            class="text-break"
          >
            [FAN0001]: Test Course 2
          </strong>
          <div
            class="card-header-btn-toolbar"
          >
            <tm-panel-chevron>
              <button
                aria-expanded="false"
                aria-label="Expand panel"
                class="chevron btn-course btn btn-primary btn-sm"
                type="button"
              >
                <i
                  class="fas fa-chevron-down"
                  style="color: white;"
                />
              </button>
            </tm-panel-chevron>
          </div>
        </div>
      </div>
    </div>
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-instructor"
      type="button"
    >
      Copy 
    </button>
    <button
      class="btn btn-light"
      id="btn-cancel-copy-instructor"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-instructors-from-other-courses-modal>
`;

exports[`CopyInstructorsFromOtherCoursesModalComponent should snap with default fields 1`] = `
<tm-copy-instructors-from-other-courses-modal
  InstructorPermissionRole={[Function Object]}
  SortBy={[Function Object]}
  SortOrder={[Function Object]}
  activeModal={[Function NgbActiveModal]}
  copyClickedEvent={[Function EventEmitter_]}
  courses={[Function Array]}
  instructorService={[Function InstructorService]}
  isCopyingSelectedInstructors="false"
  notDisplayedToStudentText={[Function String]}
  statusMessageService={[Function StatusMessageService]}
  tableComparatorService={[Function TableComparatorService]}
>
  <div
    class="modal-header"
  >
    <h4
      class="modal-title"
    >
      Copy Instructors
    </h4>
    <button
      aria-label="Close"
      class="btn-close"
      id="btn-close-modal"
      type="button"
    />
  </div><div
    class="modal-body"
    id="copy-instructor-modal"
  >
  </div><div
    class="modal-footer"
  >
    <button
      class="btn btn-primary"
      disabled=""
      id="btn-confirm-copy-instructor"
      type="button"
    >
      Copy 
    </button>
    <button
      class="btn btn-light"
      id="btn-cancel-copy-instructor"
      type="button"
    >
      Cancel
    </button>
  </div>
</tm-copy-instructors-from-other-courses-modal>
`;
